packages/arb-avm-cpp/data_storage/src/value/code.cpp

```diff
@@ -0,0 +1,266 @@
+/*
+ * Copyright 2020, Offchain Labs, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <data_storage/value/code.hpp>
+
+#include "referencecount.hpp"
+#include "utils.hpp"
+
+#include <data_storage/datastorage.hpp>
+#include <data_storage/storageresult.hpp>
+#include <data_storage/value/value.hpp>
+
+#include <avm_values/code.hpp>
+
+#include <boost/endian/conversion.hpp>
+
+#include <vector>
+
+namespace {
+
+const char* max_code_segment_key = "max_code_segment";
+constexpr auto segment_key_prefix = std::array<char, 1>{87};
+constexpr auto segment_key_size = segment_key_prefix.size() + sizeof(uint64_t);
+
+std::array<unsigned char, segment_key_size> segment_key(uint64_t segment_id) {
+    std::array<unsigned char, segment_key_size> key;
+    auto it = std::copy(segment_key_prefix.begin(), segment_key_prefix.end(),
+                        key.begin());
+    auto big_id = boost::endian::native_to_big(segment_id);
+    auto big_id_ptr = reinterpret_cast<const char*>(&big_id);
+    std::copy(big_id_ptr, big_id_ptr + sizeof(big_id), it);
+    return key;
+}
+}  // namespace
+
+void deleteCodeSegmentReferences(Transaction& transaction,
+                                 const std::vector<unsigned char>& stored_value,
+                                 std::map<uint64_t, uint64_t>& segment_counts) {
+    auto iter = stored_value.begin();
+    auto ptr = reinterpret_cast<const char*>(&*iter);
+    auto cp_count = checkpoint::utils::deserialize_uint64(ptr);
+    for (uint64_t i = 0; i < cp_count; i++) {
+        bool is_immediate = static_cast<bool>(*ptr);
+        ptr += 34;
+        if (is_immediate) {
+            uint256_t value_hash = deserializeUint256t(ptr);
+            deleteValueImpl(transaction, value_hash, segment_counts);
+        }
+    }
+}
+
+std::vector<unsigned char> prepareToSaveCodeSegment(
```

Comment:
think this function can be broken into smaller helper functions with good names so people can understand easier.

Reply:
I refactored out `serializeCodeSegment` which helps a fair amount. Do you think that's enough or does it need more breaking up?